File: Graph/3503/3503.lua
local LIMIT = 100000

local function SUM(x, y)
    return x + y
end

local function SUB(x, y)
    return x - y
end

local function MUL(x, y)
    return x * y
end

local function DIV(x, y)
    if x % y == 0 then
        return x // y
    end
    return nil
end

local MATH = {SUM, SUB, MUL, DIV}

local function BFS(source, target, nums)

    if source == target then
        return 0
    end

    local queue = {
        {value = source, operations = 0}
    }

    local visited = {}
    visited[source] = true

    local head = 1

    while head <= #queue do
        local curr = queue[head]
        head = head + 1

        for i = 1, #nums do
            local a = nums[i]

            for j = 1, 4 do
                local nextValue = MATH[j](curr.value, a)

                if nextValue
                    and nextValue >= 1
                    and nextValue <= LIMIT
                    and not visited[nextValue] then

                    if nextValue == target then
                        return curr.operations + 1
                    end

                    visited[nextValue] = true
                    queue[#queue + 1] = {
                        value = nextValue,
                        operations = curr.operations + 1
                    }
                end
            end
        end
    end

    return -1
end

local source, target, n = io.read("*n", "*n", "*n")

local nums = {}
for i = 1, n do
    nums[i] = io.read("*n")
end

print(BFS(source, target, nums))

File: Graph/__3000__/3503/3503.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <functional>
using namespace std;

const int LIMIT = 100000;

vector<function<int(int,int)>> MATHF = {
    [](int a, int b){ return a + b; }, // SUM
    [](int a, int b){ return a - b; }, // SUB
    [](int a, int b){ return a * b; }, // MUL
    [](int a, int b){                  // DIV
        if (b != 0 && a % b == 0)
            return a / b;
        return 0;
    }
};

struct State {
    int value;
    int operations;
};

int BFS(int source, int target, const vector<int>& nums){
    if (source == target) return 0;
    queue<State> q;
    unordered_set<int> vis;
    q.push({source, 0});
    vis.insert(source);

    while(!q.empty()){
        State curr = q.front();
        q.pop();
        if(curr.value == target) return curr.operations;
        for(int n : nums){
            for(auto& f : MATHF){
                int nextValue = f(curr.value, n);
                if(nextValue >= 1 && nextValue <= LIMIT && !vis.count(nextValue)){
                    vis.insert(nextValue);
                    q.push({nextValue, curr.operations + 1});
                }
            }
        }
    }
    return -1;
}

int main(){
    int source, target, n;
    cin >> source >> target >> n;
    vector<int> nums(n);
    for(int i = 0; i < n; i++)
        cin >> nums[i];

    cout << BFS(source, target, nums) << "\n";
}